### ISSUE TITLE
ENH: global optimizations that understand jac=True

### ISSUE THREAD

**Issue by tbolind** [scipy/scipy#19305]
### Is your feature request related to a problem? Please describe.

I am trying to solve high dimensional global optimization problems (500-5000 parameters) and to make this feasible I want to include jacobian information. One easy way of feeding jacobian information is by using pytorchs autograd. A method for doing this with local optimizers is described here: https://bnikolic.co.uk/blog/pytorch/python/2021/02/28/pytorch-scipyminim.html

A code example of this would be something like:
```
import numpy
import scipy, scipy.optimize
import torch


def minim(obs, f, p0):
    """Fit function f to observations "obs" starting at p0"""
    
    def fitfn(pars):
        # NB the require_grad parameter specifying we want to
        # differentiate wrt to the parameters
        pars=torch.tensor(pars,
                          requires_grad=True)
        y=f(pars)
        # Simple least-squares fitting
        res=((obs-y)**2).sum()
        res.backward()
        # Note that gradient is taken from the "pars" variable
        return res.data.cpu().numpy(), pars.grad.data.cpu().numpy()

    res=scipy.optimize.minimize(fitfn,
                                p0,
                                method="BFGS",
                                jac=True) # NB: we will compute the jacobian
    return res

# Points on which observations are made
X=torch.arange(100)

def exfn(p):
    y=torch.sin(p[0]*X)+torch.cos(p[1]*X)
    return y

# Sample observations
yp=exfn(torch.tensor([0.3, 0]) )

# Sample run
minim(yp, exfn, numpy.array([0.34, 0.01]))
```

Now I would like to use this method for global optimizers as well, but unfortunately global optimizers currently have no keyword jac.
As such they will throw an error if I try the above with a global optimizer since the global optimizer will expect my objective function to return only the value and not the gradient.

### Describe the solution you'd like.

I would like a solution that enables me to inform the global optimizer that the objective function will return both the objective value and parameter gradients, such that I can calculate the jacobian in a similar way to how it is done for a local minimizer.

### Describe alternatives you've considered.

A current workaround is that I can compute the jacobian separately from the objective evaluation. But is a less efficient approach since it will mean I have to calculate the objective function twice, rather than the natural approach where I get both at the same time.

**Comment by rkern** [comment]
AFAICT, none of our global optimizers can use gradient information, except for their final local-minimizer polishing step after the global minimizer algorithm itself is done. So computing gradient information is not going to help make an otherwise dimensionally infeasible problem tractable by these algorithms. It's not a matter of a missing API, but algorithm choice.

**Comment by andyfaff** [comment]
rkerns post is spot on. 

In addition to that... Some global optimisers (let's call them type A) use a local-minimiser for polishing at the end of the global search, other global optimisers use local-minimisers at each step (type B).

For type A (e.g. `dual_annealing`) the number of function evaluations *greatly* outnumbers the number of gradient calculations required. If you're calculating gradient at each point then there's a lot of inefficiency.

For type B (e.g. `basinhopping`) the number of function and grad evaluations are a lot closer.

There are some global minimisers that are a mix of type A and type B depending on the minimiser kwds chosen, an example of this is `shgo`.

Consequently, a lot of arguments in the OP are not True, computing FG does not lead to increased efficiency.

The following code section can be used to explore my statements. After each of the global minimises look at the number of function and gradient evaluations. If you only have a function that calculates FG, then you can use something like `MemoizeJac` to return either func or grad for the same input. You will *possibly* have a benefit for type B, but not guaranteed.

```python
import numpy as np
import jax
import jax.numpy as jnp
from scipy.optimize import dual_annealing, basinhopping, shgo
from jax import config
config.update("jax_enable_x64", True)


class MemoizeJac:
    """ Decorator that caches the return values of a function returning `(fun, grad)`
        each time it is called. """

    def __init__(self, fun):
        self.fun = fun
        self.jac = None
        self._value = None
        self.nfev = 0
        self.njev = 0
        self.x = None

    def _compute_if_needed(self, x, *args):
        if not np.all(x == self.x) or self._value is None or self.jac is None:
            self.x = np.asarray(x).copy()
            fg = self.fun(x, *args)
            self.jac = fg[1]
            self._value = fg[0]

    def __call__(self, x, *args):
        """ returns the function value """
        self._compute_if_needed(x, *args)
        self.nfev += 1
        return self._value

    def derivative(self, x, *args):
        self._compute_if_needed(x, *args)
        self.njev += 1
        return self.jac

    
def bird(x):
    return (jnp.sin(x[0]) * jnp.exp((1 - jnp.cos(x[1])) ** 2) 
            + jnp.cos(x[1]) * jnp.exp((1 - jnp.sin(x[0])) ** 2) + (x[0] - x[1]) ** 2)


dfdx = jax.grad(bird)


def fg(x):
    return np.array(bird(x)), np.array(dfdx(x))

global_optimum = np.array([[4.701055751981055, 3.152946019601391],
                            [-1.582142172055011, -3.130246799635430]])
bounds = [(-2 * np.pi, 2 * np.pi)]*2

a = fg(global_optimum[0])[0]


# type A
# massive inefficiencies for calculating FG.
memoized_fg = MemoizeJac(fg)
res = dual_annealing(
    memoized_fg,
    bounds,
    minimizer_kwargs={"jac": memoized_fg.derivative, "method":'l-bfgs-b', "bounds": bounds},
    maxiter=1000
)
print(memoized_fg.nfev, memoized_fg.njev)
print(res)
print()


# type B
# you'll note that basinhopping doesn't find the global optimum
memoized_fg = MemoizeJac(fg)
res = basinhopping(
    memoized_fg,
    x0 = [2, 2],
    minimizer_kwargs={"jac": memoized_fg.derivative, "method":'l-bfgs-b', "bounds": bounds},
)
print(memoized_fg.nfev, memoized_fg.njev)
print(res)
print()


# mix of type A and type B, depending on options.
memoized_fg = MemoizeJac(fg)
res = shgo(
    memoized_fg,
    bounds,
    minimizer_kwargs={"jac": memoized_fg.derivative, "method":'l-bfgs-b', "bounds": bounds},
    options={"jac": memoized_fg.derivative},
    sampling_method='sobol',
)
print(memoized_fg.nfev, memoized_fg.njev)
print(res)
```

gives:

```
4014 13
 message: ['Maximum number of iteration reached']
 success: True
  status: 0
     fun: -106.76453674897645
       x: [-1.582e+00 -3.130e+00]
     nit: 1000
    nfev: 4014
    njev: 13
    nhev: 0

863 863
                    message: ['requested number of basinhopping iterations completed successfully']
                    success: True
                        fun: -87.31088273300355
                          x: [-1.537e+00  3.107e+00]
                        nit: 100
      minimization_failures: 0
                       nfev: 863
                       njev: 863
 lowest_optimization_result:  message: CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL
                              success: True
                               status: 0
                                  fun: -87.31088273300355
                                    x: [-1.537e+00  3.107e+00]
                                  nit: 6
                                  jac: [-3.760e-08 -3.023e-07]
                                 nfev: 9
                                 njev: 9
                             hess_inv: <2x2 LbfgsInvHessProduct with dtype=float64>

221 93
 message: Optimization terminated successfully.
 success: True
     fun: -106.76453674926168
    funl: [-1.068e+02 -1.068e+02 -8.731e+01 -4.841e+01  1.487e+00
            1.487e+00  1.487e+00  1.487e+00]
       x: [ 4.701e+00  3.153e+00]
      xl: [[ 4.701e+00  3.153e+00]
           [-1.582e+00 -3.130e+00]
           ...
           [-5.618e+00 -5.378e+00]
           [ 6.653e-01  9.055e-01]]
     nit: 1
    nfev: 221
   nlfev: 93
   nljev: 93
   nlhev: 0
```

Therefore, I'm inclined to close the issue, as I don't think adding `jac=True` to a subset of the global minimisers will lead to efficiency. For type B a simple workaround is possible.

**Comment by tbolind** [comment]
> AFAICT, none of our global optimizer _algorithms_ can use gradient information, except for their final local-minimizer polishing step after the global minimizer algorithm itself is done. So computing gradient information is not going to help make an otherwise dimensionally infeasible problem tractable by these algorithms. It's not a matter of a missing API, but algorithm choice.

That isn't really the point I am making though. This is not a matter of whether or not the global optimizer uses the gradient information, but it is a matter of whether or not the global optimizer can understand that the objective function does not only return the objective but also the gradient. While this information might not be relevant for the global optimizer, it is very relevant for the local optimizer that runs within the global optimizer, and it is whenever these are running that jac=True saves computations.

**Comment by andyfaff** [comment]
If you are using a global optimiser that has `nfev` >> `njev` (type A) because there is not so much local minimisation, then calculation of both F and G in a `jac=True` type function (i.e. returning an F, G tuple) is wasteful because you don't need to calculate G for the most part.

In the case where the number of local minimisations is higher (type B), or you really really want to use an FG type function, then you can use the memoisation approach that I demonstrated above. Internally that is what happens in `minimize` anyway.

In my opinion modifying most of the global minimisers to accept `jac=True` is not worth it because:

1) it gives the impression that it aids computational efficiency when that's definitely not the case for Type A, which is most of the global minimisers (`differential_evolution`, `dual_annealing`, `brute`, `shgo`, `direct`). In fact it directly ruins computational efficiency.
2) In the situation where you have type B (`basinhopping`), or a final polishing step, one can use a simple memoisation outlined above to supply a callable jac with no recomputation of F/G. ie. no need for `jac=True`.
3) It complicates the logic of the global minimisers. One would have to add an additional wrapper to deal with FG functions, moreover one would have to inspect the local `minimizer_kwargs` to see `if jac is True`. It's doesn't seem logical to have to inspect local minimiser args to determine what the global minimiser objective behaviour is.

```python
def bird(x):
    return (jnp.sin(x[0]) * jnp.exp((1 - jnp.cos(x[1])) ** 2) 
            + jnp.cos(x[1]) * jnp.exp((1 - jnp.sin(x[0])) ** 2) + (x[0] - x[1]) ** 2)

def bird_np(x):
    # separate F
    return np.array(bird(x))

dfdx = jax.grad(bird)

def grad(x):
    # separate G
    return np.array(dfdx(x))

def fg(x):
    # combined FG
    return np.array(bird(x)), np.array(dfdx(x))

%timeit dual_annealing(bird_np, bounds, minimizer_kwargs={"jac": grad, "method":'l-bfgs-b', "bounds": bounds}, maxiter=1000) # calculate F/G separately
%time dual_annealing(memoized_fg, bounds, minimizer_kwargs={"jac": memoized_fg.derivative, "method":'l-bfgs-b', "bounds": bounds}, maxiter=100) # calculate FG at the same time. note only 100 iterations, and we're using time, not timeit
```

results in:
```
2.37 s ± 104 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

CPU times: user 8.68 s, sys: 84.7 ms, total: 8.77 s
Wall time: 8.86 s
```

Note that it takes 2.37 to do 1000 iterations when you're NOT calculating FG in the objective function. If you're calculating FG in the objective function then it would take 8.68 * 10 = 87 seconds to 1000 iterations. i.e. it's 36 times slower.

This is a simple function. For a complicated function the slowdown would be even worse.

**Comment by tbolind** [comment]
I agree with what you are saying that jac=True, is not the best way to go about it, but I also don't think the memoization that you are showing is the best way either.

The problem with the memoization is that it always computes the jacobian if the x value is different than the cached one, but for many algorithms I would imagine that you have some objective calculations where you do not want gradient information for while you have others where you do want it for.

So as far as I understand you have the options:

1.  Specify a separate jac function. In this case you get a free objective calculation with each jac calculation that is not used.
2. Use memoization as you suggest. In this case you cache the jacobian and only recalculate it if the x values are different. For a line search for instance this would lead to calculating expensive jacobians that are not needed.

I don't know which way would be most sensible to implement it, but it would be nice to have a third option where you perform no wasted calculations.
So essentially a third option where the optimizers can specify whether they are merely looking to evaluate the objective function or whether they are interested in the jacobian information for this x value as well. In the former case, only the objective function is calculated and in the latter both are calculated together.

However, I can understand that this might be a major reworking of the code, for something that might be a relatively small gain, and might lead to a lot of other issues, and as such is unlikely to be implemented at any point.

I will note that it might be nice to have some examples in the documentation highlighting how to easily implement automatic differentiation information in scipy.optimize and notably also in the global optimizations since this is less trivial. Especially since automatic differentiation is becoming more and more easy to use, but so far I don't think there is any official documentation or helping examples on how to utilize it in scipy?

**Comment by andyfaff** [comment]
For some local minimisers, `L-BFGS-B` is one of them, using FG type functions is more efficient.

For the type-A global minimisers the most efficient way of computation would be to have a function that solely calculates the F during the global search, but use an FG function for the local minimisation. It's more complicated to achieve this for a global minimiser because it makes sense to solely supply F as the first first argument to the minimiser, and it doesn't make sense to supply a different objective function to the local `minimizer_kwargs`.
The next best thing is to supply F to the global search and still specify a callable jac. This is the example I showed above (`dual_annealing(bird_np, bounds, minimizer_kwargs={"jac": grad, "method":'l-bfgs-b', "bounds": bounds})`).

For type-B global minimisers it may make sense to supply a memoised FG function like I demonstrated, because there are a lot more gradient calculations.

It is already straightforward to use automatic differentiation, the example I showed above demonstrates it. One simply supplies the automated grad function to the `jac` keyword.

**Comment by tbolind** [comment]
Whether they are type A or type B, I hope we can both agree that neither of the available options are optimal (since all available options will lead to some wasted calculations).  The fact that the user have to know which of these options their algorithm of choice falls into, and then adapt the way the supply the jacobian to the algorithm they select is also not ideal.

All that being said, I can see that these are not simple changes to make, but are perhaps things that should be considered for a future major refactoring of scipy.optimize if such a thing were ever to happen.

**Comment by andyfaff** [comment]
I'd like to revisit my statement above, slightly. I don't think all options lead to wasted calculations. 

Thinking a bit about the example: `dual_annealing` with a local minimizer using `L-BFGS-B` and separate F + G calculations is probably pretty optimal. I've just looked at local minimisation with L-BFGS-B. Both the following minimisations have identical `nfev` and `njev`.

```python
from scipy.optimize import rosen, rosen_der, minimize

def fg(x):
    return rosen(x), rosen_der(x)

minimize(rosen, [3,3,3], jac=rosen_der)
minimize(fg, [3, 3, 3], jac=True)
```

**Comment by tbolind** [comment]
> I'd like to revisit my statement above, slightly. I don't think all options lead to wasted calculations.
> 
> Thinking a bit about the example: `dual_annealing` with a local minimizer using `L-BFGS-B` and separate F + G calculations is probably pretty optimal. I've just looked at local minimisation with L-BFGS-B. Both the following minimisations have identical `nfev` and `njev`.
> 
> ```python
> from scipy.optimize import rosen, rosen_der, minimize
> 
> def fg(x):
>     return rosen(x), rosen_der(x)
> 
> minimize(rosen, [3,3,3], jac=rosen_der)
> minimize(fg, [3, 3, 3], jac=True)
> ```

Fair enough, so if you use specific optimizers you can get optimal performance with the current implementation, but in general I think it is fair so say that the optimization might be sub-optimal in one way or the other?

I don't think I have anything more to this topic, but I want to thank you for a very illuminating discussion, and for taking the time to give me all these thoughtful answers.
